File: src/com/intellij/idea/plugin/hybris/impex/assistance/ImpexColumnBreadcrumbsProvider.java
package com.intellij.idea.plugin.hybris.impex.assistance;

import com.intellij.idea.plugin.hybris.impex.ImpexLanguage;
import com.intellij.idea.plugin.hybris.impex.psi.ImpexAnyHeaderMode;
import com.intellij.idea.plugin.hybris.impex.psi.ImpexFullHeaderParameter;
import com.intellij.idea.plugin.hybris.impex.psi.ImpexFullHeaderType;
import com.intellij.idea.plugin.hybris.impex.psi.ImpexHeaderLine;
import com.intellij.idea.plugin.hybris.impex.psi.ImpexTypes;
import com.intellij.idea.plugin.hybris.impex.psi.ImpexValueGroup;
import com.intellij.idea.plugin.hybris.impex.utils.ImpexPsiUtils;
import com.intellij.lang.Language;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiWhiteSpace;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.ui.breadcrumbs.BreadcrumbsProvider;
import com.intellij.util.ObjectUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;

public class ImpexColumnBreadcrumbsProvider implements BreadcrumbsProvider {

    @Override
    public Language[] getLanguages() {
        return new Language[]{ImpexLanguage.getInstance()};
    }

    @Override
    public boolean acceptElement(final @NotNull PsiElement element) {
        return element instanceof ImpexFullHeaderParameter ||
               element instanceof ImpexFullHeaderType ||
               element instanceof ImpexAnyHeaderMode;
    }

    @NotNull
    @Override
    public String getElementInfo(@NotNull PsiElement psi) {
        ImpexFullHeaderParameter headerParameter = getLinkedHeaderParameter(psi);
        if (headerParameter != null) {
            return headerParameter.getText();
        }

        //ImpexHeaderLine headerLine = PsiTreeUtil.getParentOfType(element, ImpexHeaderLine.class, false);
        //assert headerLine != null;

        psi = adjustWhiteSpaceAndSeparator(psi);

        headerParameter = PsiTreeUtil.getParentOfType(psi, ImpexFullHeaderParameter.class, false);
        if (headerParameter != null) {
            return headerParameter.getText();
        }

        ImpexAnyHeaderMode mode = PsiTreeUtil.getParentOfType(psi, ImpexAnyHeaderMode.class, false);
        if (mode != null) {
            return mode.getText();
        }

        ImpexFullHeaderType type = PsiTreeUtil.getParentOfType(psi, ImpexFullHeaderType.class, false);
        if (type != null) {
            return type.getHeaderTypeName().getText();
        }

        return "<error> : " + psi.getNode().getElementType() + ": " + psi.getText();
    }

    @Nullable
    @Override
    public PsiElement getParent(@NotNull PsiElement element) {

        final ImpexFullHeaderParameter linkedParameter = getLinkedHeaderParameter(element);
        if (linkedParameter != null) {
            return linkedParameter;
        }

        final ImpexFullHeaderParameter parentParameter =
            PsiTreeUtil.getParentOfType(element, ImpexFullHeaderParameter.class, true);
        if (parentParameter != null) {
            return parentParameter;
        }

        if (element instanceof ImpexAnyHeaderMode) {
            //we are done
            return null;
        }

        PsiElement adjusted = adjustWhiteSpaceAndSeparator(element);
        if (adjusted instanceof ImpexAnyHeaderMode) {
            //we are exactly at the space just after mode, mode itself had not been yet reported
            return adjusted;
        }

        if (adjusted instanceof ImpexFullHeaderParameter) {
            ImpexHeaderLine line = PsiTreeUtil.getParentOfType(adjusted, ImpexHeaderLine.class, false);
            return Optional.ofNullable(line)
                           .map(l -> Optional.<PsiElement>ofNullable(l.getFullHeaderType()).orElse(l.getAnyHeaderMode()))
                           .orElse(line);
        }
        if (adjusted instanceof ImpexFullHeaderType) {
            ImpexHeaderLine line = PsiTreeUtil.getParentOfType(adjusted, ImpexHeaderLine.class, false);
            return line == null ? null : line.getAnyHeaderMode();
        }

        return PsiTreeUtil.getParentOfType(
            adjusted,
            ImpexValueGroup.class,
            ImpexFullHeaderParameter.class,
            ImpexFullHeaderType.class,
            ImpexAnyHeaderMode.class
        );
    }

    @Nullable
    private static ImpexFullHeaderParameter getLinkedHeaderParameter(final @NotNull PsiElement psi) {

        return Optional.ofNullable(ImpexPsiUtils.getClosestSelectedValueGroupFromTheSameLine(psi))
                       .map(ImpexPsiUtils::getHeaderForValueGroup)
                       .map(o -> ObjectUtils.tryCast(o, ImpexFullHeaderParameter.class))
                       .orElse(null);
    }

    private static PsiElement adjustWhiteSpaceAndSeparator(PsiElement psi) {
        if (psi instanceof PsiWhiteSpace) {
            //white space should be adjusted to left
            PsiElement prev = PsiTreeUtil.skipSiblingsBackward(psi, PsiWhiteSpace.class);
            if (prev != null) {
                psi = prev;
            }
        } else if (psi.getNode().getElementType() == ImpexTypes.PARAMETERS_SEPARATOR && psi.getNextSibling() != null) {
            //separator is always belongs to the *next* parameter, so adjusting to right
            psi = psi.getNextSibling();
        }
        return psi;
    }

}

File: src/com/intellij/idea/plugin/hybris/impex/assistance/ImpexColumnBreadcrumbsProvider.kt
/*
 * This file is part of "hybris integration" plugin for Intellij IDEA.
 * Copyright (C) 2019 EPAM Systems <hybrisideaplugin@epam.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.intellij.idea.plugin.hybris.impex.assistance;

import com.intellij.idea.plugin.hybris.impex.ImpexLanguage
import com.intellij.idea.plugin.hybris.impex.psi.*
import com.intellij.idea.plugin.hybris.impex.utils.ImpexPsiUtils
import com.intellij.lang.Language
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiWhiteSpace
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.ui.breadcrumbs.BreadcrumbsProvider

class ImpexColumnBreadcrumbsProvider : BreadcrumbsProvider {

    override fun getLanguages(): Array<Language> = arrayOf(ImpexLanguage.getInstance())

    override fun acceptElement(element: PsiElement) = element is ImpexFullHeaderParameter
        || element is ImpexFullHeaderType
        || element is ImpexAnyHeaderMode

    override fun getElementInfo(psi: PsiElement): String {
        val headerParameter = getLinkedHeaderParameter(psi)
        if (headerParameter != null) return headerParameter.text

        val adjustedPsi = adjustWhiteSpaceAndSeparator(psi)

        val parentParameter = PsiTreeUtil.getParentOfType(adjustedPsi, ImpexFullHeaderParameter::class.java, false)
        if (parentParameter != null) return parentParameter.text

        val mode = PsiTreeUtil.getParentOfType(adjustedPsi, ImpexAnyHeaderMode::class.java, false)
        if (mode != null) return mode.text

        val type = PsiTreeUtil.getParentOfType(adjustedPsi, ImpexFullHeaderType::class.java, false)
        if (type != null) return type.headerTypeName.text

        return "<error> : ${psi.node.elementType} : ${psi.text}"
    }

    override fun getParent(element: PsiElement): PsiElement? {
        val linkedParameter = getLinkedHeaderParameter(element)
        if (linkedParameter != null) return linkedParameter

        val parentParameter = PsiTreeUtil.getParentOfType(element, ImpexFullHeaderParameter::class.java, true)
        if (parentParameter != null) return parentParameter

        if (element is ImpexAnyHeaderMode) return null

        val adjustedPsi = adjustWhiteSpaceAndSeparator(element)
        return when (adjustedPsi) {
            is ImpexAnyHeaderMode -> adjustedPsi
            is ImpexFullHeaderParameter -> {
                val line = getImpexHeaderLine(adjustedPsi)
                line?.fullHeaderType
                    ?: line?.anyHeaderMode
            }

            is ImpexFullHeaderType -> getImpexHeaderLine(adjustedPsi)
                ?.anyHeaderMode

            else -> {
                PsiTreeUtil.getParentOfType(
                    adjustedPsi,
                    ImpexValueGroup::class.java,
                    ImpexFullHeaderParameter::class.java,
                    ImpexFullHeaderType::class.java,
                    ImpexAnyHeaderMode::class.java
                )
            }
        }
    }

    private fun getImpexHeaderLine(adjustedPsi: PsiElement) = PsiTreeUtil
        .getParentOfType(adjustedPsi, ImpexHeaderLine::class.java, false)

    private fun getLinkedHeaderParameter(psi: PsiElement): ImpexFullHeaderParameter? = ImpexPsiUtils
        .getClosestSelectedValueGroupFromTheSameLine(psi)
        ?.fullHeaderParameter

    private fun adjustWhiteSpaceAndSeparator(psiElement: PsiElement): PsiElement {
        if (psiElement is PsiWhiteSpace) {
            val previousElement = PsiTreeUtil.skipSiblingsBackward(psiElement, PsiWhiteSpace::class.java)
            if (previousElement != null) return previousElement
        } else if (isParameterSeparator(psiElement)) {
            return psiElement.nextSibling
        }
        return psiElement
    }

    private fun isParameterSeparator(psi: PsiElement) = psi.node.elementType == ImpexTypes.PARAMETERS_SEPARATOR
        && psi.nextSibling != null
}